jax/_src/third_party/scipy/special.py

```diff
@@ -0,0 +1,324 @@
+  - https://github.com/scipy/scipy/blob/v1.14.0/scipy/special/special/cephes/trig.h
+  """
+  x = jnp.abs(x)
+  # define s = x % 2, y = x - s, then
+  # r = (x * x / 2) % 2
+  #   = [(y + s)*(y + s)/2] % 2
+  #   = [y*y/2 + s*y + s*s/2] % 2
+  #   = [(y*y/2)%2 + (s*y + s*s/2)%2]%2
+  #   = [0 + (s*(y+s/2))%2]%2
+  #   = [s*(x-s/2)]%2
+  s = jnp.fmod(x, 2.0)
+  r = jnp.fmod(s * (x - s / 2), 2.0)
+
+  sinpi = jnp.where(
+    r < 0.5,
+    jnp.sin(jnp.pi * r),
+    jnp.where(
+      r > 1.5,
+      jnp.sin(jnp.pi * (r - 2.0)),
+      -jnp.sin(jnp.pi * (r - 1.0)),
+    ),
+  )
+  cospi = jnp.where(
+    r == 0.5,
+    0.0,
+    jnp.where(r < 1.0, -jnp.sin(jnp.pi * (r - 0.5)), jnp.sin(jnp.pi * (r - 1.5))),
+  )
+
+  return sinpi, cospi
+
+
+@custom_derivatives.custom_jvp
+def fresnel(x: ArrayLike) -> tuple[Array, Array]:
+  r"""The Fresnel integrals
+
+  JAX implementation of :obj:`scipy.special.fresnel`.
+
+  The Fresnel integrals are defined as
+    .. math::
+       S(x) &= \int_0^x \sin(\pi t^2 /2) dt \\
+       C(x) &= \int_0^x \cos(\pi t^2 /2) dt.
+
+  Args:
+    x: arraylike, real-valued.
+
+  Returns:
+    Arrays containing the values of the Fresnel integrals.
+
+  Notes:
+     The JAX version only supports real-valued inputs, and
+     is based on the SciPy C++ implementation, see
+     `here
+     <https://github.com/scipy/scipy/blob/v1.14.0/scipy/special/special/cephes/fresnl.h>`_.
+     For ``float32`` dtypes, the implementation is directly based
+     on the Cephes implementation ``fresnlf``.
+
+     As for the original Cephes implementation, the accuracy
+     is only guaranteed in the domain [-10, 10]. Outside of
+     that domain, one could observe divergence between the
+     theoretical derivatives and the custom JVP implementation,
+     especially for large input values.
+
+     Finally, for half-precision data types, ``float16``
+     and ``bfloat16``, the array elements are upcasted to
+     ``float32`` as the Cephes coefficients used in
+     series expansions would otherwise lead to poor results.
+     Other data types, like ``float8``, are not supported.
+  """
+  if jnp.iscomplexobj(x):
+    raise NotImplementedError(
+        'Support for complex-valued inputs is not implemented yet.')
+
+  xxa, = promote_args_inexact("fresnel", x)
+
+  # This part is mostly a direct translation of SciPy's C++ code,
+  # and the original Cephes implementation for single precision.
+
+  orig_dtype = xxa.dtype
+
+  if orig_dtype in (jnp.float16, jnp.bfloat16):
+    # For half-precision, series expansions have either
+    # produce overflow or poor accuracy.
+    # Upcasting to single-precision is hence needed.
+    dtype = jnp.float32
+    xxa = xxa.astype(dtype)
+  elif orig_dtype in (jnp.float32, jnp.float64):
+    dtype = orig_dtype
+  else:
+    raise NotImplementedError(
+        f'Support for {orig_dtype} dtype is not implemented yet.')
+
+  single_precision = (dtype == jnp.float32)
+
+  if single_precision:
```

Comment:
This may seem nitpicky, but handling the logic of dtype support with two different if-else blocks will make this code harder to maintain in the long run. A future reader of this code trying to debug an issue with, say, a new dtype, will have to parse by eye several dozen lines of code & understand the provenance of several related variables (`single_precision`, `dtype`, and `original_dtype`) in order to understand how the logic depends on the input dtype. It would be much more immediately understandable to put all this logic in one place. For example:
```python
if dtypes.issubdtype(xxa.dtype, dtypes.complexfloating):
  raise ValueError("complex values not supported")
elif xxa.dtype == jnp.float64:
  # define double-precision parameters
elif xxa.dtype in (jnp.float32, jnp.float16, jnp.bfloat16):
  xxa = xxa.astype('float32')
  # define single-precision paramters
else:
  raise ValueError(f"unsuppored dtype: {dtype}")

assert xxa.dtype in ('float32', 'float64')
single_precision = (xxa.dtype == 'float32')
```
What do you think?

Reply:
I agree this could be simpler, what do you think of d69866043e9731275405ac1eb33811c3ad843bb5 ?